semeio/workflows/ahm_analysis/ahmanalysis.py

```diff
@@ -0,0 +1,456 @@
+    )
+    ESUpdate(ert).smootherUpdate(run_context)
+
+
+def _get_field_params(ert, ensemble_size, field_parameters, target_name):
+    """
+    Because the FIELD parameters are not exposed in the Python API we have to
+    export them to file and read them back again. When they are exposed in the API
+    this function should be updated.
+    """
+    field_data = {}
+    file_system = ert.getEnkfFsManager().getFileSystem(target_name)
+    with tempfile.TemporaryDirectory() as fout:
+        for field_param in field_parameters:
+            config_node = ert.ensembleConfig()[field_param]
+            ext = config_node.get_enkf_outfile().rsplit(".")[-1]
+            file_path = os.path.join(fout, "%d_" + field_param + "." + ext)
+            _export_field_param(config_node, file_system, ensemble_size, file_path)
+            fnames = glob.glob(os.path.join(fout, "*" + field_param + "*"))
+            field_data[field_param] = _import_field_param(
+                ert.eclConfig().get_gridfile(), field_param, fnames
+            )
+    return field_data
+
+
+def _import_field_param(input_grid, param_name, files):
+    mygrid_char = Grid(input_grid.rsplit(".", 1)[0], fformat="eclipserun")
+    all_input = []
+    for file_path in files:
+        proproff = GridProperty(file_path, name=param_name, grid=mygrid_char)
+        array_nb = proproff.get_npvalues1d(activeonly=False, fill_value=0, order="C")
+        all_input.append(array_nb)
+    return all_input
+
+
+def _export_field_param(config_node, file_system, ensemble_size, output_path):
+    # Get/export the updated Field parameters
+    EnkfNode.exportMany(
+        config_node,
+        output_path,
+        file_system,
+        np.arange(0, ensemble_size),
+    )
+
+
+def initialize_emptydf():
+    ks_data = pd.DataFrame()
+    active_obs = pd.DataFrame()
+    misfitval = pd.DataFrame()
+    return ks_data, active_obs, misfitval
+
+
+def make_update_log_df(update_log_file, key):
+    """ Read update_log file to get active and inactive observations """
+    list_of_files = [
+        os.path.join(update_log_file, f) for f in os.listdir(update_log_file)
+    ]
+    if len(list_of_files) > 1:
+        raise OSError("ERROR more than one update_log_file.")
+    # read file
+    updatelog = pd.read_csv(
+        list_of_files[0],
+        delim_whitespace=True,
+        skiprows=6,
+        usecols=[0, 2, 3, 5, 6, 8, 10],
+        header=None,
+    )
+    # define header
+    updatelog.columns = [
+        "obs_number",
+        "obs_key",
+        "obs_mean",
+        "obs_std",
+        "status",
+        "sim_mean",
+        "sim_std",
+    ]
+    # drop last line ('=====...====')
+    # drop last row i.e. end of file line
+    updatelog.drop(updatelog.tail(1).index, inplace=True)
+    # ---------------------------------
+    # add proper name for obs_key in rows where value is '...'
+    replace = "..."
+    for i in range(len(updatelog.index)):
```

Comment:
Maybe I'm misunderstanding this, but the test data is fairly simple, as it only contains the single observation. With an input of more than a single observation, all the `...` gets the same value, e.g. using `allobs` test data:
```
  36  : RWI_3_OBS                                285.000 +/-            3.000  Active   |           273.958 +/-          10.529
  37  :   ...                                    258.000 +/-            3.000  Active   |           264.119 +/-           9.031
  38  :   ...                                    271.000 +/-            3.000  Active   |           269.818 +/-           5.951
  39  : RWI_2_OBS                                291.000 +/-            3.000  Active   |           272.672 +/-          14.870
  40  :   ...                                    233.000 +/-            3.000  Active   |           265.408 +/-          14.494
  41  :   ...                                    297.000 +/-            3.000  Active   |           277.500 +/-          13.479
  42  : ROP_5_OBS                                285.000 +/-            3.000  Active   |           274.673 +/-           6.869
  43  :   ...                                    279.000 +/-            3.000  Active   |           271.760 +/-           5.229
```

```
5          36    RWI_3_OBS    285.00     3.00    Active   273.958   10.529
36          37    RWI_3_OBS    258.00     3.00    Active   264.119    9.031
37          38    RWI_3_OBS    271.00     3.00    Active   269.818    5.951
38          39    RWI_2_OBS    291.00     3.00    Active   272.672   14.870
39          40    RWI_3_OBS    233.00     3.00    Active   265.408   14.494
40          41    RWI_3_OBS    297.00     3.00    Active   277.500   13.479
```

Is it only made to handle single observations in the updagelog perhaps?

Reply:
Here's a suggestion to use pandas `ffill` instead.

```
    df = pd.read_csv(
        list_of_files[0],
        delim_whitespace=True,
        skiprows=6,
        usecols=[0, 2, 3, 5, 6, 8, 10],
        header=None,
        skipfooter=1,
    )
    df.columns = [
        "obs_number",
        "obs_key",
        "obs_mean",
        "obs_std",
        "status",
        "sim_mean",
        "sim_std",
    ]
    # Identical values for obs_key will have the obs_value in first record
    # and subsequently "..." for the following records. We want an exhaustive list
    df.replace("...", np.nan, inplace=True) # ffill requires np.nan
    df.ffill(inplace=True)

    return df
```

Reply:
ffill implemented (was not initially since there was an issue in updatelog: when the first point in gan_data had no simulated data/point outside the grid for instance then the first data would be ... so not possible to know to which well it belonged, but this is now fixed so ffill works perfect now)